corehq/apps/domain/tests/test_auth.py

```diff
@@ -64,3 +72,252 @@ def test_request_for_valid_domain_membership_for_non_superuser(self, *args):
+
+    def test_user_can_authenticate_if_ip_is_on_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+        request = self._create_request(username='test@dimagi.com', api_key='1234', ip='127.0.0.1')
+
+        auth = HQApiKeyAuthentication()
+
+        self.assertTrue(auth.is_authenticated(request))
+
+    def test_user_cannot_authenticate_with_an_ip_not_in_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+        request = self._create_request(username='test@dimagi.com', api_key='1234', ip='5.5.5.5')
+
+        auth = HQApiKeyAuthentication()
+
+        response = auth.is_authenticated(request)
+
+        self.assertEqual(response.status_code, 401)
+
+    def test_user_can_authenticate_with_current_key(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', expiration=datetime(year=2020, month=3, day=10))
+        request = self._create_request(username='test@dimagi.com', api_key='1234')
+
+        auth = HQApiKeyAuthentication()
+
+        with freeze_time(datetime(year=2020, month=3, day=9)):
+            self.assertTrue(auth.is_authenticated(request))
+
+    def test_user_cannot_authenticate_with_expired_key(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', expiration=datetime(year=2020, month=3, day=10))
+        request = self._create_request(username='test@dimagi.com', api_key='1234')
+
+        auth = HQApiKeyAuthentication()
+
+        with freeze_time(datetime(year=2020, month=3, day=11)):
+            response = auth.is_authenticated(request)
+            self.assertEqual(response.status_code, 401)
+
+    @classmethod
+    def _create_user(cls, username):
+        return User.objects.create_user(username=username, password='password')
+
+    @classmethod
+    def _create_api_key_for_user(
+            cls,
+            user,
+            name='ApiKey',
+            key='1234',
+            is_active=True,
+            allowed_ips=None,
+            domain='',
+            expiration=None
+    ):
+        allowed_ips = allowed_ips or []
+        return HQApiKey.objects.create(
+            user=user,
+            key=key,
+            name=name,
+            ip_allowlist=allowed_ips,
+            domain=domain,
+            is_active=is_active,
+            expiration_date=expiration
+        )
+
+    @classmethod
+    def _create_request(cls, username='test@dimagi.com', api_key='1234', for_domain='test-domain', ip=None):
+        request = HttpRequest()
+        request.domain = for_domain
+        request.META['HTTP_AUTHORIZATION'] = f'ApiKey {username}:{api_key}'
+
+        if ip:
+            request.META['HTTP_X_FORWARDED_FOR'] = ip
+
+        return request
```

Comment:
These functions are duplicated across two test classes. What do you think of moving them to module-level functions so both tests can use the same helpers?

Reply:
The main method that is important here is `_create_request`, and that is not duplicated, as the two test suites need different ways of providing authentication (Fallback is likely receiving BASIC authentication, whereas `HQApiKey` only touches `ApiKey` authentication).

`_create_user` and `_create_api_key_for_user` are technically one-liner helper functions, and their duplication, I think, strongly hints that the common functionality between the two modules-under-test should be refactored, rather than creating common utility functions between two test suites that are basically the same.